compliance/collection/nodeinventorizer/caching_inventorizer.go

```diff
@@ -0,0 +1,284 @@
+	return &CachingScanner{
+		inventoryCachePath:  inventoryCachePath,
+		cacheDuration:       cacheDuration,
+		initialBackoff:      initialBackoff,
+		maxBackoff:          maxBackoff,
+		backoffWaitCallback: backoffCallback,
+	}
+}
+
+// Scan scans the current node and returns the results as storage.NodeInventory struct.
+// A cached version is returned if it exists and is fresh enough.
+// Otherwise, a new scan guarded by a backoff is run.
+func (c *CachingScanner) Scan(nodeName string) (*storage.NodeInventory, error) {
+	// check whether a cached inventory exists that has not exceeded its validity
+	cache := readInventoryWrap(c.inventoryCachePath)
+	if cache != nil && cache.Inventory != nil && cache.ValidUntil.After(time.Now()) {
+		log.Debugf("Using cached node scan (valid until %v)", cache.ValidUntil)
+		return cache.Inventory, nil
+	}
+
+	// check for existing backoff, wait for specified duration if needed, then persist the new backoff duration
+	backoffDuration := c.initialBackoff
+	if cache != nil {
+		backoffDuration = c.validateBackoff(cache.BackoffDuration)
+	}
+
+	if backoffDuration > c.initialBackoff {
+		log.Warnf("Found existing node scan backoff file - last scan may have failed. Waiting %v seconds before retrying", backoffDuration.Seconds())
+		c.backoffWaitCallback(backoffDuration)
+	}
+
+	// Write backoff duration to cache
+	backoff := inventoryWrap{BackoffDuration: c.calcNextBackoff(backoffDuration)}
+	if err := writeInventoryWrap(backoff, c.inventoryCachePath); err != nil {
+		log.Warnf("Error writing node scan backoff file: %v", err)
+	}
+
+	// if no inventory exists, or it is too old, collect a fresh one
+	newInventory, err := collectInventory(nodeName)
+	if err != nil {
+		return nil, err
+	}
+
+	// Write inventory to cache
+	inventory := inventoryWrap{
+		ValidUntil:      time.Now().Add(c.cacheDuration),
+		BackoffDuration: 0,
+		Inventory:       newInventory,
+	}
+	if err := writeInventoryWrap(inventory, c.inventoryCachePath); err != nil {
+		return nil, errors.Wrap(err, "persisting inventory to cache")
+	}
+
+	return newInventory, nil
+}
+
+func (c *CachingScanner) calcNextBackoff(currentBackoff time.Duration) time.Duration {
+	nextBackoffInterval := currentBackoff * backoffMultiplier
+	if nextBackoffInterval > c.maxBackoff {
+		return c.maxBackoff
+	}
+	return nextBackoffInterval
+}
+
+// validateBackoff ensures that a given duration does not exceed the max backoff setting
+func (c *CachingScanner) validateBackoff(backoff time.Duration) time.Duration {
+	if backoff > c.maxBackoff {
+		return c.maxBackoff
+	}
+	return backoff
+}
+
+func readInventoryWrap(path string) *inventoryWrap {
+	cacheContents, err := os.ReadFile(path)
+	if err != nil {
+		if errors.Is(err, os.ErrNotExist) {
+			log.Debug("No node scan cache file found, will run a new scan")
```

Comment:
I think, "will run a new scan" thingy now becomes a bit out of place. I suggest simply removing it here and in other log statements in this function.
```suggestion
			log.Debug("No node scan cache file found")
```